compiler/rustc_codegen_ssa/src/back/apple.rs

```diff
@@ -26,6 +26,87 @@ pub(super) fn macho_platform(target: &Target) -> u32 {
     }
 }
 
+/// Add relocation and section data needed for a symbol to be considered
+/// undefined by ld64.
+///
+/// Inherently very architecture-specific (unfortunately).
+///
+/// # New architectures
+///
+/// The values here can be found by compiling the following program:
+///
+/// ```c
+/// void foo(void);
+///
+/// void use_foo() {
+///     foo();
+/// }
+/// ```
+///
+/// With:
+///
+/// ```console
+/// $ clang -c foo.c -O2 -g0 -fno-exceptions -target $CLANG_TARGET
+/// ```
+///
+/// And then inspecting with `objdump -d foo.o` and/or:
+///
+/// ```rust,ignore
+/// //! ```cargo
+/// //! [dependencies]
+/// //! object = "0.36"
+/// //! ```
+///
+/// use object::read::macho::{MachHeader, MachOFile};
+/// use object::{File, Object, ObjectSection};
+///
+/// fn read(file: MachOFile<'_, impl MachHeader>) {
+///     for section in file.sections() {
+///         dbg!(section.name().unwrap(), section.data().unwrap(), section.align());
+///         for reloc in section.relocations() {
+///             dbg!(reloc);
+///         }
+///     }
+/// }
+///
+/// fn main() {
+///     match File::parse(&*std::fs::read("foo.o").unwrap()).unwrap() {
+///         File::MachO32(file) => read(file),
+///         File::MachO64(file) => read(file),
+///         _ => unimplemented!(),
+///     }
+/// }
+/// ```
+pub(super) fn add_data_and_relocation(
+    file: &mut object::write::Object<'_>,
+    section: object::write::SectionId,
+    symbol: object::write::SymbolId,
+    target: &Target,
+) -> object::write::Result<()> {
+    let (data, align, addend, r_type): (&[u8], _, _, _) = match &*target.arch {
+        "arm" => (&[0xff, 0xf7, 0xfe, 0xbf], 2, 0, object::macho::ARM_THUMB_RELOC_BR22),
+        "aarch64" => (&[0, 0, 0, 0x14], 4, 0, object::macho::ARM64_RELOC_BRANCH26),
+        "x86_64" => (
+            &[0x55, 0x48, 0x89, 0xe5, 0x5d, 0xe9, 0, 0, 0, 0],
+            16,
+            -4,
+            object::macho::X86_64_RELOC_BRANCH,
+        ),
+        "x86" => (&[0x55, 0x89, 0xe5, 0x5d, 0xe9, 0xf7, 0xff, 0xff, 0xff], 16, -4, 0),
+        arch => unimplemented!("unsupported Apple architecture {arch:?}"),
+    };
```

Comment:
I'm quite unhappy with this, it _feels_ very brittle to emit a manual relocation like this, but I think it's the only way to do it (without pulling in a bunch of LLVM codegen internals).

I'm also unsure of the correctness, there might be more target options and/or flags that we should be checking here? Or maybe we should emit different kinds of relocations based on the kind of symbol?